src/eval/cache.rs

```diff
@@ -19,13 +20,25 @@ pub trait Cache: Clone {
     fn get_then<T, F: FnOnce(&Closure) -> T>(&self, idx: CacheIndex, f: F) -> T;
     fn update(&mut self, clos: Closure, idx: Self::UpdateIndex);
     fn new() -> Self;
-    fn reset_index_state(&self, idx: &mut Self::UpdateIndex);
+    fn reset_index_state(&mut self, idx: &mut Self::UpdateIndex);
     fn map_at_index<F: FnMut(&Closure) -> Closure>(&mut self, idx: &CacheIndex, f: F)
         -> CacheIndex;
-    // TODO: Needs a better name
     fn build_cached(&mut self, idx: &mut CacheIndex, rec_env: &[(Ident, CacheIndex)]);
+    fn ident_kind(&self, idx: CacheIndex) -> IdentKind;
+    fn saturate<'a, I: DoubleEndedIterator<Item = &'a Ident> + Clone>(
+        &mut self,
+        idx: CacheIndex,
+        env: &mut Environment,
+        fields: I,
+    ) -> RichTerm;
+    fn revert(&mut self, idx: CacheIndex) -> CacheIndex;
```

Comment:
`revert` is taking a immutable reference in `Thunk`. Is there any reason why it needs an owned variant here? I know in a future ideal world `CacheIndex` will be `Copy`, probably. But right now we take an immutable reference e.g. for `map_at_index`. It think it would make sense to take an immutable reference as well for `ident_kind`, `revert`, and `deps`, which don't mutate anything. That would get rid of some thunk cloning.

Reply:
I left these from the `DumbCache` interface, which used `usize` as indices for the cache and resorted to cloning thunks around just to make it respect the interface since we talked about having something easily copy-able as indices in the future, but I'm not really attached to this choice, I'll change it